rllib/models/base_model.py

```diff
@@ -0,0 +1,350 @@
+# Copyright 2021 DeepMind Technologies Limited.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import abc
+from typing import Optional, Tuple
+
+from ray.rllib.models.temp_spec_classes import TensorDict, SpecDict, ModelConfig
+from ray.rllib.utils.annotations import (
+    DeveloperAPI,
+    OverrideToImplementCustomLogic,
+    PublicAPI,
+)
+
+
+ForwardOutputType = TensorDict
+# [Output, Recurrent State(s)]
+UnrollOutputType = Tuple[TensorDict, TensorDict]
+
+
+class RecurrentModel(abc.ABC):
+    """The base model all other models are based on.
+
+    RLlib models all inherit from the recurrent base class, which can be chained
+    together with other models.
+
+    The models input and output TensorDicts. Which keys the models read/write to
+    and the desired tensor shapes must be defined in input_spec, output_spec,
+    prev_state_spec, and next_state_spec.
+
+    The `unroll` function gets the model inputs and previous recurrent state, and
+    outputs the model outputs and next recurrent state. Note all ins/outs must match
+    the specs. Users should override `_unroll` rather than `unroll`.
+
+    `initial_state` returns the "next" state for the first recurrent iteration. Again,
+    users should override `_initial_state` instead.
+
+    For non-recurrent models, users may instead override `_forward` which does not
+    make use of recurrent states.
+
+    Args:
+        name: An optional name for the module
+
+    Examples:
+        >>> named_class = NamedClass() # Inherits from RecurrentModel
```

Comment:
I'm confused on what examples we would use for an abstract class. Your example would not work if the user typed it in the shell, because the shape methods and `_unroll` are not overridden.

Reply:
For abstract classes it won't be runnable. But it should show how the class will be used one layer above.